### PR TITLE
feat: Move Campers Modal

### PR DESCRIPTION
**Pull request by uyxela**
## Notion ticket link

[Camp Overview: Move Camper(s) to different session](https://www.notion.so/uwblueprintexecs/Task-Board-F22-9e03bf9f7c4343f09a8a5c16dc48a12a?p=16ca401098364231a674e581ea836899&pm=c)

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/implementations/camperService.ts [hunk 1]
```diff
@@ -602,6 +605,318 @@ class CamperService implements ICamperService {
...
+        charges: updatedCamper.charges,
+        optionalClauses: updatedCamper.optionalClauses,
+      }));
+    } catch (error: unknown) {
+      await session.abortTransaction();
+      await session.endSession();
+      throw error;
+    }
+
+    // let updatedCamperDTOs: Array<CamperDTO> = [];
```
Comment (dhmi19) [inline backend/typescript/services/implementations/camperService.ts:734]: Do we still need these comments? Can we delete for cleanup?

File: backend/typescript/services/implementations/camperService.ts [hunk 2]
```diff
@@ -288,12 +289,14 @@ class CamperService implements ICamperService {
           lastName: camper.lastName,
           age: camper.age,
           allergies: camper.allergies,
-          earlyDropoff: camper.earlyDropoff.map((date) => date.toString()),
-          latePickup: camper.latePickup.map((date) => date.toString()),
+          earlyDropoff: camper.earlyDropoff.map((date) => date?.toString()),
```
Comment (dhmi19) [inline backend/typescript/services/implementations/camperService.ts:292]: Did you make these changes? There seem to be a bunch of changes that are not related to your ticket?

File: frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx [hunk 3]
```diff
@@ -125,6 +149,34 @@ const CampersTable = ({
     setCamperDetailsCount(tempDetailsCount);
   }, [campers]);
 
+  // Function to be called by the Move Campers Modal to update the campers
```
Comment (Jasraj-Dhanoa) [inline frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx:152]: Alert alignment should be more refined. It tends to hide when scrolling up and overlaps the delete button. I noticed that the error alert was not showing any message whenever the camper could not be moved. Another way to make the interaction better would be to reduce the time it's displayed on the page. Also, is there a specific reason for choosing the alert compared to using a toast (which also provides the user the functionality to close it whenever they like)?

Screen Shot 2022-10-21 at 10 22 47 PM
Screen Shot 2022-10-21 at 10 34 24 PM

**Review by Jasraj-Dhanoa — CHANGES_REQUESTED** [review]
I tested out this feature with campID 63139c7bc3d7b55b44a01531

The functionality for Scenario 1 and Scenario 2 were working as expected. However, I was receiving an error when trying to test Scenario 3 (the alert didn't specify what specific error was occurring). It may be due to inconsistency within the database, but it will be great if you could take a look and test it out. Also, I've added some comments addressing the alert functionality and styling.

Please let me know if you have any questions.

File: backend/typescript/services/implementations/camperService.ts [hunk 4]
```diff
@@ -602,6 +605,318 @@ class CamperService implements ICamperService {
     return updatedCamperDTOs;
   }
 
+  async moveCampersToCampSession(
+    camperIds: string[],
+    campSessionId: string,
+  ): Promise<CamperDTO[]> {
+    const session = await mongoose.startSession();
+    await session.startTransaction();
```
Comment (dhmi19) [inline backend/typescript/services/implementations/camperService.ts:613]: is this await necessary? I got a warning saying this is not necessary and the function documentation doesn't mention anything about Promises

File: frontend/src/components/pages/CampOverview/MoveCampersModal.tsx [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+                marginLeft: "10px",
+              }}
+              onClick={() => {
+                setModalPage(2);
+              }}
+              disabled={
+                camperIsSelected.reduce<number>(
+                  (prev, cur) => (cur ? prev + 1 : prev),
+                  0,
+                ) === 0
```
Comment (dhmi19) [inline frontend/src/components/pages/CampOverview/MoveCampersModal.tsx:151]: An admin can choose not to move any grouped campers so this condition is not accurate. I don't think the next button can ever be disabled for this page

File: frontend/src/components/pages/CampOverview/MoveCampersModal.tsx [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+              You can move the camper(s) to another session in this camp by
+              selecting it below:
+            </p>
+            {campSessions.length > 0 ? (
+              <Select
+                value={selectedCampSession}
+                onChange={(e) => setSelectedCampSession(e.target.value)}
+              >
+                {campSessions.map((campSessionsItem, campSessionsIdx) => (
+                  <option key={campSessionsIdx} value={campSessionsItem.id}>
```
Comment (dhmi19) [inline frontend/src/components/pages/CampOverview/MoveCampersModal.tsx:183]: we need to "disable" the option if the camp session is the current session or a session in the past. We also need to disable the session if it doesn't have enough spots available. 
image

File: frontend/src/components/pages/CampOverview/MoveCampersModal.tsx [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+                    setCamperIsSelected(
+                      Array<boolean>(campers.length).fill(false),
+                    );
+                  }
+                }}
+              >
+                Select all
+              </Checkbox>
+              {campers.map((campersItem, i) => (
+                <Checkbox
```
[the remark this answers is not in the document]
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/MoveCampersModal.tsx:116]: Can you also remove the current selected camper from this list? eg: Tejas Test 1 should not be in this list

File: frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx [hunk 8]
```diff
@@ -74,6 +93,11 @@ const CampersTable = ({
     onClose: viewModalOnClose,
   } = useDisclosure();
 
+  const [alertInfo, setAlertInfo] = React.useState<{
```
[the remark this answers is not in the document]
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx:96]: image
The toast should also disappear after 3 seconds and only take as much width as the text needs (or as specified by figma)

**Review by dhmi19 — CHANGES_REQUESTED** [review]

File: frontend/src/components/pages/CampOverview/MoveCamperModal/index.tsx [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+  // Check if the camper has group members
+  useEffect(() => {
+    // Used to prevent the flash of previous state
+    reset();
+
+    // Check if the camper object is not null
+    if (Object.keys(camper).length !== 0) {
+      // Check if camper has group members
+      CamperAPIClient.getCampersByChargeId(camper.chargeId).then(
```
Comment (dhmi19) [inline frontend/src/components/pages/CampOverview/MoveCamperModal/index.tsx:49]: Sorry about the confusion, this should be `getCampersByChargeIdAndSessionId`. We should only be able to move campers who were registered together and who are part of the current selected session. So you would pass in the current session id to the sessionID parameter.

File: frontend/src/components/pages/CampOverview/MoveCamperModal/PageTwo.tsx [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+          <Select
+            value={selectedCampSession}
+            onChange={(e) => setSelectedCampSession(e.target.value)}
+          >
+            {campSessions.map((campSessionsItem, campSessionsIdx) => (
+              <option
+                key={campSessionsIdx}
+                value={campSessionsItem.id}
+                disabled={
+                  campSessionsItem.id === camper.campSession ||
```
Comment (dhmi19) [inline frontend/src/components/pages/CampOverview/MoveCamperModal/PageTwo.tsx:60]: I think you missed a check here to disable all sessions which have an end date before today's date. IE: admin shouldn't be able to select a session in the past (session which has completed)

**Review by dhmi19 — CHANGES_REQUESTED** [review]
coming along well, just 2 final changes from my end I think. Sorry about the confusion with moving campers from the particular session :(

**Comment by twilkhoo** [comment]
Completed here https://github.com/uwblueprint/focus-on-nature/pull/247.

**Pull request closed without merging**